### PR TITLE
Added Test for RatingIdentifierAlreadyExist

### LINKED ISSUE #298: [C4GT] Network-Score: Add tests for 'RatingIdentifierAlreadyAdded'
## Description
Subtask under : https://github.com/cord-network/community/issues/7

New testcase can be tested under `cargo test -p pallet-network-score` after adding it in the code.

## Goals

- [ ] Add tests for `RatingIdentifierAlreadyAdded` for pallet/network-score

## Expected Outcome
- [ ] Test should assert for `RatingIdentifierAlreadyAdded` being returned properly in all the possible calls.
- [ ] If there are more than one function returning this error code, all of them should be validated either as one test case or as multiple test cases.
       
## Acceptance Criteria
NA

## Implementation Details
Look at other test cases, and add a test case for the same.
Ex: In pallets/network-membership check method `test_duplicate_member_request` which validates error code `MembershipAlreadyAcquired`

## Mockups / Wireframes
NA

---- 

## Product Name
CORD

## Organization Name
Dhiway

## Domain
Blockchain

## Tech Skills Needed
Rust

## Mentor(s)
@amarts 

## Complexity
[Low]

## Category
[Test]

## Sub Category
[Beginner friendly]

### PR DESCRIPTION
**Pull request by zeel991**
- Fixes https://github.com/dhiway/cord/issues/298
- In this test ,When the Registered Rating is used again , it throws the error `RatingIdentifierAlreadyExist` 
- The `revise_rating()` method expects the `EntryTypeOf` to be Debit , So `revoke_rating()` method has been called .

### CODE REVIEW

**Review by vatsa287 — COMMENTED** [review]
Reviews from other PR #422 
Revise rating expects entry type to be of Debit in the storage. To test this can be done by

Calling revoke_rating which does exactly this ( Ideal ).
Setting entry-type to be of DEBIT in storage ( Not Recommended in this case since we can do through a method from above (revoke_rating()).
You can try again, I'll post entire solution tomorrow.

**Comment by zeel991** [comment]
Hey @vatsa287 
I tried Using the `revise_rating()` method but it gave `ReferenceNotDebitIdentifier` Error 
```
assert_err!(
  Score::revoke_rating(
	DoubleOrigin(author.clone(), creator.clone()).into(),
	identifier.clone(),
	message_id.clone(),
	entry_digest,
	authorization_id.clone(),
  ),
   Error::<Test>::RatingIdentifierAlreadyAdded
);
```
Can You Please review it

**Comment by zeel991** [comment]
Can we use the `update_activity()` method to change the `EntryTypeOf` to `Debit` ?

**Comment by zeel991** [comment]
Hello @vatsa287 , Can you guide me on how to successfully revoke a rating ?

File: pallets/network-score/src/tests.rs [hunk 1]
```diff
@@ -384,3 +384,91 @@ fn revoke_rating_with_existing_rating_identifier_should_fail() {
...
+		<RatingEntries<Test>>::insert(
+			&identifier,
+			RatingEntryOf::<Test> {
+				creator_id: author.clone(),
+				entry_type: EntryTypeOf::Debit,
+				reference_id: Some(identifier.clone()),
+				..rating_details
+			},
+		);
+
```
Comment (vatsa287) [inline pallets/network-score/src/tests.rs:499]: revoke_rating will take the identifier returned (created during) register_rating, with optional new messgae_id.
When `revoke_rating` happens it creates a `debit` entry in the chain which will be useful for give a new rating through `revise_rating` to create a new `credit` entry.

Comment (zeel991) [reply pallets/network-score/src/tests.rs:499]: Ohk ! @vatsa287  , So i have to get the rating which was created during `register_rating` , but in my current approach , i have generated an identifier which is created from the arguments of `register_rating` function , why is that giving `RatingIdentifierAlreadyAdded` error ?

Comment (vatsa287) [reply pallets/network-score/src/tests.rs:499]: Please paste the screenshot with the code. I'll look up and review.

Comment (zeel991) [reply pallets/network-score/src/tests.rs:499]:
```
let id_digest = <Test as frame_system::Config>::Hashing::hash(
&[
	&entry_digest.encode()[..],
	&entry.entity_id.encode()[..],
	&message_id.encode()[..],
	&space_id.encode()[..],
	&creator.clone().encode()[..],
]
.concat()[..],);
let identifier =
       Ss58Identifier::create_identifier(&(id_digest).encode()[..], IdentifierType::Rating)
	.unwrap();
```
Screenshot 2024-04-15 162927

**Comment by vatsa287** [comment]
@zeel991 Do the following,

```rust
new_test_ext().execute_with(|| {
        System::set_block_number(1);

        // Create a space
        assert_ok!(Space::create(
            DoubleOrigin(author.clone(), creator.clone()).into(),
            space_digest
        ));
        assert_ok!(Space::approve(RawOrigin::Root.into(), space_id, 5u64));

        // Register the rating entry once
        assert_ok!(Score::register_rating(
            DoubleOrigin(author.clone(), creator.clone()).into(),
            entry.clone(),
            entry_digest,
            message_id_add.clone(),
            authorization_id.clone(),
        ));
	
	// Revoke the rating to create a debit entry
        assert_ok!(
            Score::revoke_rating(
                DoubleOrigin(author.clone(), creator.clone()).into(),
                identifier_add.clone(),
                message_id_revoke.clone(),
                entry_digest.clone(),
                authorization_id.clone()
            )
        );
	
	// Revise rating to create a new rating for the entity
        assert_ok!(
            Score::revise_rating(
                DoubleOrigin(author.clone(), creator.clone()).into(),
                entry_revise.clone(),
                entry_revise_digest.clone(),
                message_id_revise.clone(),
                identifier_revoke.clone(),
                authorization_id.clone(),
            ),
        );
	
	// Remove the messgae_id from list to reach `RatingIdentifierAlreadyAdded` block.
        <MessageIdentifiers<Test>>::remove(message_id_revise.clone(), creator.clone());
	
	// Error when revising rating again with same rating identifier
        assert_err!(
            Score::revise_rating(
                DoubleOrigin(author.clone(), creator.clone()).into(),
                entry_revise.clone(),
                entry_revise_digest.clone(),
                message_id_revise.clone(),
                identifier_revoke,
                authorization_id.clone(),
            ),
            Error::<Test>::RatingIdentifierAlreadyAdded
        );
    });
```

**Comment by zeel991** [comment]
> @zeel991 Do the following,
> 
> ```rust
> new_test_ext().execute_with(|| {
>         System::set_block_number(1);
> 
>         // Create a space
>         assert_ok!(Space::create(
>             DoubleOrigin(author.clone(), creator.clone()).into(),
>             space_digest
>         ));
>         assert_ok!(Space::approve(RawOrigin::Root.into(), space_id, 5u64));
> 
>         // Register the rating entry once
>         assert_ok!(Score::register_rating(
>             DoubleOrigin(author.clone(), creator.clone()).into(),
>             entry.clone(),
>             entry_digest,
>             message_id_add.clone(),
>             authorization_id.clone(),
>         ));
> 	
> 	// Revoke the rating to create a debit entry
>         assert_ok!(
>             Score::revoke_rating(
>                 DoubleOrigin(author.clone(), creator.clone()).into(),
>                 identifier_add.clone(),
>                 message_id_revoke.clone(),
>                 entry_digest.clone(),
>                 authorization_id.clone()
>             )
>         );
> 	
> 	// Revise rating to create a new rating for the entity
>         assert_ok!(
>             Score::revise_rating(
>                 DoubleOrigin(author.clone(), creator.clone()).into(),
>                 entry_revise.clone(),
>                 entry_revise_digest.clone(),
>                 message_id_revise.clone(),
>                 identifier_revoke.clone(),
>                 authorization_id.clone(),
>             ),
>         );
> 	
> 	// Remove the messgae_id from list to reach `RatingIdentifierAlreadyAdded` block.
>         <MessageIdentifiers<Test>>::remove(message_id_revise.clone(), creator.clone());
> 	
> 	// Error when revising rating again with same rating identifier
>         assert_err!(
>             Score::revise_rating(
>                 DoubleOrigin(author.clone(), creator.clone()).into(),
>                 entry_revise.clone(),
>                 entry_revise_digest.clone(),
>                 message_id_revise.clone(),
>                 identifier_revoke,
>                 authorization_id.clone(),
>             ),
>             Error::<Test>::RatingIdentifierAlreadyAdded
>         );
>     });
> ```

ok

File: pallets/network-score/src/tests.rs [hunk 2]
```diff
@@ -384,3 +384,130 @@ fn revoke_rating_with_existing_rating_identifier_should_fail() {
 		);
 	});
 }
+
+#[test]
+fn test_revise_rating_id_already_exists() {
```
Comment (vatsa287) [inline pallets/network-score/src/tests.rs]: nit: Follow other naming conventions as in,
`fn revise_rating_with_existing_rating_identifier_should_fail()`

**Comment by zeel991** [comment]
Hey @vatsa287 , Please review the PR

**Review by vatsa287 — APPROVED** [review]
LGTM

**Comment by zeel991** [comment]
> LGTM

Thank you so much , @amarts , Please review the PR

**Review by amarts — APPROVED** [review]

**Comment by zeel991** [comment]
Hey @vatsa287 , the PR is approved , can you please merge it

**Pull request merged**